pkg/utils/capacityclient/client.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2025 The KubeFleet Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package capacityclient
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"google.golang.org/protobuf/encoding/protojson"
+
+	computev1 "go.goms.io/fleet/pkg/protos/azure/compute/v1"
+)
+
+const (
+	// recommendationsPathTemplate is the URL path template for VM size recommendations API.
+	recommendationsPathTemplate = "/subscriptions/%s/providers/Microsoft.Compute/locations/%s/vmSizeRecommendations/vmAttributeBased/generate"
+)
+
+// CapacityClientFactory is a function type for creating CapacityClient instances.
+type CapacityClientFactory func(endpoint string, httpClient *http.Client) CapacityClient
+
+// CapacityClient is an interface for interacting with the Azure Capacity API.
+type CapacityClient interface {
+	// GenerateAttributeBasedRecommendations generates VM size recommendations based on attributes.
+	GenerateAttributeBasedRecommendations(ctx context.Context, req *computev1.GenerateAttributeBasedRecommendationsRequest) (*computev1.GenerateAttributeBasedRecommendationsResponse, error)
+}
+
+var _ CapacityClient = &client{}
+
+// client implements the CapacityClient interface for interacting with Azure Capacity API.
+type client struct {
+	// baseURL is the base URL of the capacity service endpoint.
+	baseURL string
+	// httpClient is the HTTP client used for making requests.
+	httpClient *http.Client
+}
+
+// NewClient creates a new capacity client with the given endpoint and HTTP client.
+// If httpClient is nil, http.DefaultClient will be used.
+// If the endpoint does not have a scheme (http:// or https://), like localhost:8080, http:// will be added.
+func NewClient(endpoint string, httpClient *http.Client) CapacityClient {
+	if httpClient == nil {
+		httpClient = http.DefaultClient
```

Comment:
Using http.DefaultClient is discouraged in production,

https://vishnubharathi.codes/blog/know-when-to-break-up-with-go-http-defaultclient/

Reply:
added timeout.